src/adapter/src/client.rs

```diff
@@ -246,6 +274,21 @@ pub struct SessionClient {
 }
 
 impl SessionClient {
+    /// Take the session from the client so it is not `Drop`ed, causing a
+    /// Terminate request which will fail if the connection startup failed.
+    pub fn take_session(mut self) {
+        // TODO(mjibson): Requiring this for error'd connections is a bad API because
+        // it's invisible and will only manifest on failed connections, and so could
+        // easily be forgotten during development. Improve the API to make it not
+        // misusable.
```

Comment:
Hrm, yeah, this API seems treacherous. Is it possible that perhaps it's `Session::new` that should generate the secret key, and store it in the session? Then I don't think you'd need to change the `SessionClient` API! You'd have the secret key available exactly where you need it in the pgwire protocol, and it would get plumbed through via the session into all the right places in the coordinator.